src/geos_types.jl

```diff
@@ -323,51 +348,78 @@ end
+function compare(cmp, geo1::AbstractGeometry, geo2::AbstractGeometry, ctx=get_context(geo1))::Bool
     (typeof(geo1) === typeof(geo2)) || return false
-    ng1 = ngeom(geo1)
-    ng2 = ngeom(geo2)
-    ng1 == ng2 || return false
-    for i in 1:ng1
-        (getgeom(geo1, i) ≅ getgeom(geo2, i)) || return false
+    if (geo1 === geo2) && (cmp === isequal)
+        return true
     end
-    return true
-end
-function compare(≅, pt1::Point, pt2::Point)::Bool
-    is3d = GeoInterface.is3d(pt1)
-    is3d === GeoInterface.is3d(pt2) || return false
-    GeoInterface.getcoord(pt1,1) ≅ GeoInterface.getcoord(pt2,1) || return false
-    GeoInterface.getcoord(pt1,2) ≅ GeoInterface.getcoord(pt2,2) || return false
-    if is3d
-        GeoInterface.getcoord(pt1,3) ≅ GeoInterface.getcoord(pt2,3) || return false
+    if has_coord_seq(geo1)
+        return compare_coord_seqs(cmp, geo1, geo2, ctx)
+    else
+        ng1 = ngeom(geo1)
+        ng2 = ngeom(geo2)
+        ng1 == ng2 || return false
+        for i in 1:ng1
+            compare(cmp, getgeom(geo1, i), getgeom(geo2, i), ctx) || return false
+        end
     end
     return true
 end
 
-function _norm(x,y,z)
-    return sqrt(x^2 + y^2 + z^2)
+npoints(pt::Point) = 1
+npoints(geo::HasCoordSeq) = GeoInterface.ngeom(geo)
+
+function compare_coord_seqs(cmp, geo1, geo2, ctx)
+    ncoords1 = GeoInterface.ncoord(geo1)
+    ncoords2 = GeoInterface.ncoord(geo2)
+    ncoords1 == ncoords2 || return false
+    ncoords1 == 0 && return true
+    np1 = npoints(geo1)
+    np2 = npoints(geo2)
+    np1 == np2 || return false
+    coords1 = Vector{Float64}(undef, ncoords1)
+    coords2 = Vector{Float64}(undef, ncoords1)
+    for i in 1:np1
+        coordinates!(coords1, geo1, i, ctx)
+        coordinates!(coords2, geo2, i, ctx)
+        cmp(coords1, coords2) || return false
+    end
+    return true
 end
 
-function compare(cmp::IsApprox, pt1::Point, pt2::Point)::Bool
-    is3d = GeoInterface.is3d(pt1)
-    is3d === GeoInterface.is3d(pt2) || return false
-    x1 = GeoInterface.getcoord(pt1,1)
-    x2 = GeoInterface.getcoord(pt2,1)
-    y1 = GeoInterface.getcoord(pt1,2)
-    y2 = GeoInterface.getcoord(pt2,2)
-    if is3d
-        z1 = GeoInterface.getcoord(pt1,3) 
-        z2 = GeoInterface.getcoord(pt2,3)
-    else
-        z1 = 0.0
-        z2 = 0.0
+function compare_coord_seqs(cmp::IsApprox, geo1, geo2, ctx)
+    ncoords1 = GeoInterface.ncoord(geo1)
+    ncoords2 = GeoInterface.ncoord(geo2)
+    ncoords1 == ncoords2 || return false
+    ncoords1 == 0 && return true
+    @assert ncoords1 in 2:3
+    ncoords1 == ncoords2 || return false
+    np1 = npoints(geo1)
+    np2 = npoints(geo2)
+    np1 == np2 || return false
+    coords1 = Vector{Float64}(undef, ncoords1)
+    coords2 = Vector{Float64}(undef, ncoords1)
+    # isapprox of two vectors is calculated using their euclidean norms and the norm of their difference
+    # we compute (the squares) of these norms in an allocation free way
```

Comment:
(optional) out of curiosity -- how important is it to be using the euclidean norm v.s. the L1 norm? I.e.

```julia
s12 += abs(x2 - x2) + abs(y1 - y2) [+ abs(z1 - z2)]
# ...
return s12 <= cmp.atol # + ...
```

Reply:
We could use another norm, I choose L2 because it is what `norm` computes by default.

Reply:
And `norm` is used in `isapprox` by default.

Reply:
Thank you!